### PR TITLE
code optimisation for circular ref issue

### PR DESCRIPTION
**Pull request by SB-rohitdesai**
Mock server not working In case of circular references. 

## Description
Getting 500 Error or Sometimes might need to wait a long time before getting this error when spec having circular references

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] This PR's code follows as closely as possible [the coding style/guidelines of this project](???).
- [ ] I have added error reporting and followed [the error reporting guidelines](???).
- [ ] I have added event tracking and followed [the event tracking guidelines](???).
- [ ] I have updated any relevant documentation accordingly to reflect this PR's changes.
- [ ] I have added automated tests (unit/integration/e2e/other) to cover my changes.
- [ ] All new and existing tests pass locally (excluding flaky CI tests).

### CODE REVIEW

**Review by rainum — COMMENTED** [review]
@SB-rohitdesai, it looks good! Do you plan to add any tests for this functionality though?

**Comment by SB-rohitdesai** [comment]
> @SB-rohitdesai, it looks good! Do you plan to add any tests for this functionality though?

Actually everything is already covered to test async await functionality.

File: lib/dereference.js [hunk 1]
```diff
@@ -34,18 +42,34 @@ function dereference (parser, options) {
...
+  parents,
+  dereferencedCache,
+  $refs,
+  options
+) {
   let dereferenced;
   let result = {
     value: obj,
-    circular: false
+    circular: false,
```
Comment (bhaskarsontakke) [inline lib/dereference.js]: remove the comma on line no. 57.

**Review by bhaskarsontakke — APPROVED** [review]
Changes looks good to me. Though few comments are added

File: lib/dereference.js [hunk 2]
```diff
@@ -165,7 +225,7 @@ function dereference$Ref ($ref, path, pathFromRoot, parents, dereferencedCache,
 
   const dereferencedObject = {
     circular,
-    value: dereferencedValue
+    value: dereferencedValue,
```
Comment (bhaskarsontakke) [inline lib/dereference.js]: same here remove comma

File: lib/dereference.js (hunk 1 above)
Comment (SB-rohitdesai) [reply lib/dereference.js]: Done

File: lib/dereference.js (hunk 2 above)
Comment (SB-rohitdesai) [reply lib/dereference.js]: Done

**Review by rainum — COMMENTED** [review]
All those functions look synchronous to me 🤔 Why did you make them async? Can we get some benchmarking results on how it improves the derefencing speed?

**Comment by SB-rohitdesai** [comment]
> All those functions look synchronous to me 🤔 Why did you make them async? Can we get some benchmarking results on how it improves the derefencing speed?

Hi @rainum 
As I updated all three files (In all three PR's) , we are able to reduce the time , if you want to see that , you can check the video which I attached to the ticket

**Comment by SB-rohitdesai** [comment]
> > All those functions look synchronous to me 🤔 Why did you make them async? Can we get some benchmarking results on how it improves the derefencing speed?
> 
> Hi @rainum As I updated all three files (In all three PR's) , we are able to reduce the time , if you want to see that , you can check the video which I attached to the ticket

Hi @rainum May be because of async its faster , But its giving response quickly than previous code

**Comment by rainum** [comment]
> > > All those functions look synchronous to me 🤔 Why did you make them async? Can we get some benchmarking results on how it improves the derefencing speed?
> > 
> > 
> > Hi @rainum As I updated all three files (In all three PR's) , we are able to reduce the time , if you want to see that , you can check the video which I attached to the ticket
> 
> Hi @rainum May be because of async its faster , But its giving response quickly than previous code

We can't solve issues by using an "imperial" approach. There should be a reason for this. I don't want to have so kind of "magical" code in our codebase.

@brendarearden, do you have any thoughts regarding this one?

**Comment by brendarearden** [comment]
As far as I know making a function async shouldn't do anything to the execution time of the individual function and it might slightly increase the time because it has to create a promise in addition to running the function. So there is definitely a different underlying reason as to why this change optimized the execution time and I agree that we should have an answer to that. Is it possible that an await was not added somewhere so we just never wait for it to execute? I feel like the linter would've caught that.

**Comment by brendarearden** [comment]
I'd recommend using yalc or something similar to pull these changes into Prism, create a draft PR with the yalced version and add tests in the prism draft pr to demonstrate the optimization if you are not able to add test directly in here.

**Pull request closed without merging**